#161545388 Migrate swagger documentation to postman

davenjeru opened the pull request:
#### What does this PR do?
Migrates swagger documentation to Postman.

#### Description of Task to be completed?
 - Create postman collections.
 - Publish the said collections
 - Add 'Run in Postman' button on README

#### How should this be manually tested?
1. Git clone the repository.
2. Checkout to `ch-swagger-to-postman-161545388`.
3. Install necessary dependencies.
  `npm install` _(or the `yarn` equivalent)_.
4. Ensure the `.env` file is updated with the required variables.
5. Start the server `npm run start:dev` _(or the `yarn` equivalent)_.
6. On the README, click the `Run in Postman` button and it will automatically download and import the collections.
7. You can test the API endpoints and create a mock server on the imported collections.

#### Any background context you want to provide?
You should already have installed Postman.
You _may_ need a Postman account.
The collections' environment  variable `baseUrl` has been set to `http://localhost:8000`, you can change this to whatever host and port that the application has started on.

#### What are the relevant pivotal tracker stories?
* [#161545388](https://www.pivotaltracker.com/story/show/161545388)

#### Postman Documentation
Click this [link](https://documenter.getpostman.com/view/3235336/S11HteM1).

#### Screenshots (If applicable)
![image](https://user-images.githubusercontent.com/41262970/52480693-7dbf1980-2bbd-11e9-927e-62eaa180c79e.png)

nzediegwu1 commented:
You said: `Create a method to generate a swagger.json file from the swagger schema which
  can now be imported into Postman.`. I am curious to know what the swagger schema is all about. Is it updated by a developer or generated automatically?

samuelkubai commented on server/app.js:
```diff
@@ -0,0 +1,250 @@
+    // then call this function recursively with it as the first argument,
+    // it's handle as the second argument, and the given options as the third
+    if (layer.name === 'router') {
+      const regexPath = layer.regexp.toString().split('\\')[1];
```
@davenjeru why are we splitting instead of adding the path to the defined path directly?

davenjeru replied:
The layer, if it was defined using `route.use()` or if it was given an `express.Router` as its handler, will not have a path but a regex that will define the path. We split to get an array of strings and the path will be defined on the second element. Although I haven't tested it yet with paths that need params like `:id`.

samuelkubai commented on server/app.js:
```diff
@@ -0,0 +1,250 @@
+    // to them and the route that they handle, so we add this definition to the swagger
+    // schema for each method
+    return loopSequentially(Object.keys(layer.route.methods), (routerMethod) => {
+      const path = `${definedPath}${layer.route.path}`;
```
@davenjeru using the name `path` here may conflict with the imported path module

samuelkubai replied:
@davenjeru I think you would rather rename the import since you have used the variable `path` in a lot of places.

davenjeru replied:
Okay...

samuelkubai commented on server/app.js:
```diff
@@ -0,0 +1,250 @@
+  // newInstanceMethods with our own function that passes the arguments to our swagger
+  // handler which will then handle the addition of the paths, if any, to the swagger schema
+  await loopSequentially(Object.keys(newInstanceMethods), (method) => {
+    if (method) {
```
@davenjeru when would no method be provided here?

davenjeru replied:
The 'View' method is given by postman as a HTTP method but there is no method like `app.view` for an express app.

samuelkubai commented on server/app.js:
```diff
@@ -0,0 +1,250 @@
+
+  // Here set the swagger docs then handle 404s
+  // It shows an example of how you can add swagger doc specification using an options object.
+  app.use('/api-docs', swaggerUi.serve, swaggerUi.setup(app.schema), {
```
@davenjeru why is this being done both in the `app.js` file and `index.js` file?

davenjeru replied:
So, The app schema changes in the callback for app.listen in app.js when defining the servers, so it needs the new app.schema to serve it.
It is also defined in index.js just before the `app.use(*)` because if we don't, `/api-docs` will be handled by the handler for `app.use(*)` and we don't want that.

samuelkubai commented:
@davenjeru is there a way I can get the values I require to setup this application locally?? Feel free to say NO if there's some security issues or something.

cc. @psibomana

nzediegwu1 commented:
Nice job. Could you fix the failing test

davenjeru commented:
Having read the acceptance criteria, I decided that the best route to go was to document the API using Postman collections and share by adding the badge on the README. Automatically generating swagger.json proved to be more complicated than I anticipated, therefore, to save time and deliver on more pressing tasks, I decided to go the Postman way. The swagger generator has been exported as a [private external project](https://github.com/davenjeru/express-swagger-generator). If anyone would like to contribute, you could request for access.

nzediegwu1 commented:
Great job! Could you please add documentation with examples for these andela external api's for the sake of backend devs who would be making use of them:

1. https://api-prod.andela.com/api/v1/partners // get partners
2. https://api-prod.andela.com/api/v1/placements
*queries*
```
status=Placed - Awaiting Onboarding
status=External Engagements - Rolling Off

davenjeru commented:
Okay.

davenjeru commented:
@nzediegwu1, the `/placements/` endpoint is protected. I don't think a dev would be able to access it because for me it returns a 403. However, there is an endpoint `mock-api/placements` exposed by our product which is documented.
I have added the documentation and examples for `/partners` under the `External APIs` folder

nzediegwu1 commented:
> @nzediegwu1, the `/placements/` endpoint is protected. I don't think a dev would be able to access it because for me it returns a 403. However, there is an endpoint `mock-api/placements` exposed by our product which is documented.
> I have added the documentation and examples for `/partners` under the `External APIs` folder

Okay then, just include examples for mock-api documentation for the following queries:
```
status=Placed - Awaiting Onboarding
status=External Engagements - Rolling Off
```

davenjeru commented:
> External Engagements - Rolling Off

I've added the examples for these URL params although the only difference is the `status` key. I've also added a description for this.

davenjeru commented:
cc @psibomana

davenjeru commented on server/jobs/onboarding/freckle.js:
```diff
@@ -9,7 +9,7 @@ import { creatOrUpdatePartnerRecord } from '../../modules/automations';
  * @returns {undefined}
  */
-export default async (placement) => {
+const freckleOnboarding = async (placement) => {
```
This small change of naming and then exporting the function is only meant to increase coverage as a file that was previously covered has been deleted and was dropping the coverage. This change compensates for the drop in coverage. cc @valmayaki.

valmayaki replied:
Thanks
